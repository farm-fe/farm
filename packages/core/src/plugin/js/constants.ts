export const VITE_DEFAULT_ASSETS = [
  // images
  'apng',
  'png',
  'jpe?g',
  'jfif',
  'pjpeg',
  'pjp',
  'gif',
  'svg',
  'ico',
  'webp',
  'avif',

  // media
  'mp4',
  'webm',
  'ogg',
  'mp3',
  'wav',
  'flac',
  'aac',
  'opus',
  'mov',
  'm4a',
  'vtt',

  // fonts
  'woff2?',
  'eot',
  'ttf',
  'otf',

  // other
  'webmanifest',
  'pdf',
  'txt'
];
// the name of the virtual module internal the adapter
export const VITE_ADAPTER_VIRTUAL_MODULE = 'vite-adapter-virtual:';
