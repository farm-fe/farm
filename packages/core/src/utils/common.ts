import readline from 'node:readline';

export function clearScreen() {
  const repeatCount = process.stdout.rows - 2;
  const blank = repeatCount > 0 ? '\n'.repeat(repeatCount) : '';
  console.log(blank);
  readline.cursorTo(process.stdout, 0, 0);
  readline.clearScreenDown(process.stdout);
}

export function isObject(value: unknown): value is Record<string, unknown> {
  return Object.prototype.toString.call(value) === '[object Object]';
}

export function isEmptyObject<T extends object>(obj: T): boolean {
  return Reflect.ownKeys(obj).length === 0;
}

// Taken from https://github.com/sindresorhus/slash/blob/main/index.js (MIT)
export function slash(path: string) {
  const isExtendedLengthPath = /^\\\\\?\\/.test(path);
  // eslint-disable-next-line no-control-regex
  const hasNonAscii = /[^\u0000-\u0080]+/.test(path);

  if (isExtendedLengthPath || hasNonAscii) {
    return path;
  }

  return path.replace(/\\/g, '/');
}
