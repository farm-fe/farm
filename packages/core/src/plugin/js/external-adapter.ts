import { CUSTOM_KEYS } from '../../config/constants.js';
import {
  mergeCustomExternal,
  partialExternal
} from '../../config/normalize-config/normalize-external.js';
import { UserConfig } from '../../config/types.js';
import { isArray } from '../../utils/share.js';
import { JsPlugin } from '../type.js';

/**
 * avoid add new external in config hook
 */
export function externalAdapter(): JsPlugin {
  return {
    name: 'farm:external-adapter',

    priority: -Infinity,

    config(config: UserConfig): UserConfig | Promise<UserConfig> {
      if (
        config?.compilation?.external &&
        isArray(config.compilation.external)
      ) {
        let [stringExternal, recordExternal] = mergeCustomExternal(
          config?.compilation,
          partialExternal(config.compilation.external)
        );

        return {
          compilation: {
            external: stringExternal,
            custom: {
              [CUSTOM_KEYS.external_record]: JSON.stringify(recordExternal)
            }
          }
        };
      }
    }
  };
}
