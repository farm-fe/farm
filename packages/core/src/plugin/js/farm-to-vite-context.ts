import type { PluginContext } from 'rollup';
import type { UserConfig } from '../../config/types.js';
import type { CompilationContext } from '../type.js';
import { relative } from 'path';
import { DefaultLogger } from '../../utils/logger.js';

const contextCache = new Map<string, PluginContext>();

export function farmContextToViteContext(
  farmContext: CompilationContext,
  currentHandlingFile?: string,
  pluginName?: string,
  hookName?: string,
  config?: UserConfig
): PluginContext {
  const cacheKey = pluginName + hookName;
  if (contextCache.has(cacheKey)) {
    return contextCache.get(cacheKey) as PluginContext;
  }
  const logger = new DefaultLogger();

  const log = (message: any) => {
    if (typeof message === 'function') {
      message = message();
    }

    console.log(message);
  };

  const cacheError = () => {
    throw new Error(
      `Vite plugin ${pluginName} is not compatible with Farm for now. Because cache(called by hook ${pluginName}.${hookName}) is not supported in Farm`
    );
  };

  const viteContext: PluginContext = {
    addWatchFile: (id) => {
      if (!currentHandlingFile) {
        throw new Error(
          `Vite plugin ${pluginName} is not compatible with Farm for now. Because addWatchFile(called by hook ${pluginName}.${hookName}) can only be called in load hook or transform hook in Farm.`
        );
      }
      farmContext.addWatchFile(currentHandlingFile, id);
    },
    debug: log,
    emitFile: (params) => {
      if (params.type === 'asset') {
        let content: number[] = [];

        if (typeof params.source === 'string') {
          content = [...Buffer.from(params.source)];
        } else {
          content = [...params.source];
        }

        farmContext.emitFile({
          resolvedPath: currentHandlingFile ?? 'vite-plugin-adapter',
          name: params.fileName ?? params.name,
          content,
          resourceType: 'asset'
        });

        return 'vite-plugin-adapter-unsupported-reference-id';
      } else {
        throw new Error(
          `Vite plugin ${pluginName} is not compatible with Farm for now. Because emitFile(called by hook ${pluginName}.${hookName}) can only emit asset in Farm.`
        );
      }
    },
    error: (message): never => {
      if (typeof message === 'object') {
        farmContext.error(JSON.stringify(message));
      } else {
        farmContext.error(message);
      }

      return undefined as unknown as never;
    },
    getFileName: () => {
      throw new Error(
        `Vite plugin ${pluginName} is not compatible with Farm for now. Because getFileName(called by hook ${pluginName}.${hookName}) is not supported in Farm`
      );
    },
    getModuleIds: () => {
      throw new Error(
        `Vite plugin ${pluginName} is not compatible with Farm for now. Because getModuleIds(called by hook ${pluginName}.${hookName}) is not supported in Farm`
      );
    },
    getModuleInfo: () => {
      throw new Error(
        `Vite plugin ${pluginName} is not compatible with Farm for now. Because getModuleInfo(called by hook ${pluginName}.${hookName}) is not supported in Farm`
      );
    },
    getWatchFiles: () => {
      return farmContext.getWatchFiles();
    },
    info: log,
    load: (_) => {
      throw new Error(
        `Vite plugin ${pluginName} is not compatible with Farm for now. Because load(called by hook ${pluginName}.${hookName}) is not supported in Farm`
      );
    },
    meta: {
      rollupVersion: '3.29.4',
      watchMode: config.compilation.mode !== 'production'
    },
    parse: (_) => {
      throw new Error(
        `Vite plugin ${pluginName} is not compatible with Farm for now. Because parse(called by hook ${pluginName}.${hookName}) is not supported in Farm`
      );
    },
    resolve: async (source, importer, options) => {
      if (options.custom.caller === `${pluginName}.${hookName}`) {
        return null;
      }

      const farmResolveResult = await farmContext.resolve(
        {
          source,
          importer: {
            relativePath: relative(config.compilation.root, importer),
            queryString: importer.split('?')[1] ?? ''
          },
          kind: options.isEntry ? 'entry' : 'import'
        },
        {
          meta: {},
          caller: `${pluginName}.${hookName}`
        }
      );

      if (farmResolveResult) {
        return {
          id: farmResolveResult.resolvedPath,
          external: farmResolveResult.external,
          resolvedBy: 'vite-plugin-adapter-farm-resolve',
          moduleSideEffects: farmResolveResult.sideEffects,
          meta: {
            ...farmResolveResult.meta,
            caller: `${pluginName}.${hookName}`
          },
          // TODO these 2 options are not supported in farm
          assertions: {},
          syntheticNamedExports: false
        };
      }

      return null;
    },
    setAssetSource(assetReferenceId, source) {
      this.emitFile({
        type: 'asset',
        source,
        name: assetReferenceId
      });
    },
    warn: (message) => {
      if (typeof message === 'object') {
        farmContext.warn(JSON.stringify(message));
      } else if (typeof message === 'function') {
        farmContext.warn(JSON.stringify(message()));
      } else {
        farmContext.warn(message);
      }
    },
    cache: {
      set: cacheError,
      get: cacheError,
      delete: cacheError,
      has: cacheError
    },
    moduleIds: new Set<string>()[Symbol.iterator](),
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore Vite specific property
    getCombinedSourcemap() {
      logger.warn(
        '`vite-plugin-adapter`: getCombinedSourcemap is not supported in Farm for now. It will always return undefined.'
      );
      return undefined;
    }
  };

  contextCache.set(cacheKey, viteContext);

  return viteContext;
}
