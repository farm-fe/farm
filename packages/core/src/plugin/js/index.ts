import merge from 'lodash.merge';
import { Config } from '../../../binding/index.js';
import {
  type JsPlugin,
  normalizeDevServerOptions,
  type UserConfig
} from '../../index.js';
import {
  DEFAULT_FILTERS,
  getCssModuleType,
  VITE_PLUGIN_DEFAULT_MODULE_TYPE
} from './utils.js';
import { VitePluginAdapter } from './vite-plugin-adapter.js';
import { existsSync, readFileSync } from 'node:fs';

// export * from './jsPluginAdapter.js';
export { VitePluginAdapter } from './vite-plugin-adapter.js';

type VitePluginType = object | (() => { vitePlugin: any; filters: string[] });
type VitePluginsType = VitePluginType[];

export function handleVitePlugins(
  vitePlugins: VitePluginsType,
  userConfig: UserConfig,
  finalConfig: Config['config']
): JsPlugin[] {
  const jsPlugins: JsPlugin[] = [];

  if (vitePlugins.length) {
    userConfig = merge({}, userConfig, {
      compilation: finalConfig,
      server: normalizeDevServerOptions(userConfig.server, finalConfig.mode)
    });
  }

  for (const vitePluginObj of vitePlugins) {
    let vitePlugin = vitePluginObj,
      filters = DEFAULT_FILTERS;

    if (typeof vitePluginObj === 'function') {
      const { vitePlugin: plugin, filters: f } = vitePluginObj();
      vitePlugin = plugin;
      filters = f;
    }
    processVitePlugin(vitePlugin, userConfig, filters, jsPlugins);
  }

  // if vitePlugins is not empty, append a load plugin to load file
  // this plugin is only for compatibility
  if (vitePlugins.length) {
    jsPlugins.push({
      name: 'farm:load',
      // has to be the last one
      priority: -100,
      load: {
        filters: {
          resolvedPaths: DEFAULT_FILTERS
        },
        executor: async (params) => {
          const { resolvedPath } = params;

          // skip lazy compiled module and non-exist file
          if (
            VitePluginAdapter.isFarmInternalVirtualModule(resolvedPath) ||
            !existsSync(resolvedPath)
          ) {
            return null;
          }

          const content = readFileSync(resolvedPath, 'utf-8');

          return {
            content,
            moduleType: VITE_PLUGIN_DEFAULT_MODULE_TYPE
          };
        }
      },
      transform: {
        filters: {
          resolvedPaths: DEFAULT_FILTERS,
          moduleTypes: []
        },
        executor: async (params) => {
          const { content, moduleId, moduleType, resolvedPath } = params;

          // skip lazy compiled module and non-exist file
          if (VitePluginAdapter.isFarmInternalVirtualModule(resolvedPath)) {
            return null;
          }
          const cssModules = finalConfig?.css?.modules?.paths ?? [
            '\\.module\\.(css|less|sass|scss)$'
          ];
          // skip css module because it will be handled by Farm
          const isCssModules = cssModules.some((reg) =>
            new RegExp(reg).test(moduleId)
          );

          // treat all scss/less/.etc lang as css
          // plugin should handle css module by itself
          if (getCssModuleType(moduleId) && !isCssModules) {
            return {
              content,
              moduleType: 'css'
            };
          }

          return {
            content,
            moduleType
          };
        }
      }
    });
  }

  return jsPlugins;
}

export function processVitePlugin(
  vitePlugin: VitePluginType,
  userConfig: UserConfig,
  filters: string[],
  jsPlugins: JsPlugin[]
) {
  const processPlugin = (plugin: any) => {
    const vitePluginAdapter = new VitePluginAdapter(
      plugin as any,
      userConfig,
      filters
    );
    convertPlugin(vitePluginAdapter);
    jsPlugins.push(vitePluginAdapter);
  };

  if (Array.isArray(vitePlugin)) {
    vitePlugin.forEach((plugin) => processPlugin(plugin));
  } else {
    processPlugin(vitePlugin);
  }
}

export function convertPlugin(plugin: JsPlugin): void {
  if (
    plugin.transform &&
    !plugin.transform.filters?.moduleTypes &&
    !plugin.transform.filters?.resolvedPaths
  ) {
    throw new Error(
      `transform hook of plugin ${plugin.name} must have at least one filter(like moduleTypes or resolvedPaths)`
    );
  }
  if (plugin.transform) {
    if (!plugin.transform.filters.moduleTypes) {
      plugin.transform.filters.moduleTypes = [];
    } else if (!plugin.transform.filters.resolvedPaths) {
      plugin.transform.filters.resolvedPaths = [];
    }
  }

  if (plugin.load?.filters?.resolvedPaths?.length) {
    if (process.platform === 'win32') {
      // replace / to \
      plugin.load.filters.resolvedPaths = plugin.load.filters.resolvedPaths.map(
        (item) => item.replaceAll('/', '\\\\')
      );
    }
  }

  if (plugin.transform?.filters?.resolvedPaths?.length) {
    if (process.platform === 'win32') {
      // replace / to \
      plugin.transform.filters.resolvedPaths =
        plugin.transform.filters.resolvedPaths.map((item) =>
          item.replaceAll('/', '\\\\')
        );
    }
  }
}
