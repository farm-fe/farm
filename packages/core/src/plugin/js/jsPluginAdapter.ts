import { toArray } from '../../utils/index.js';

export function adaptorVitePlugin<UserOptions = Record<string, never>>(
  factory: any
) {
  return (userOptions?: UserOptions) => {
    const meta: any = {
      framework: 'vite'
    };
    const rawPlugins = toArray(factory(userOptions!, meta));

    const plugins = rawPlugins.map((rawPlugin: any) => {
      const plugin = transformFarmPlugin(rawPlugin);

      return plugin;
    });

    return plugins.length === 1 ? plugins[0] : plugins;
  };
}

function transformFarmPlugin(plugin: any) {
  console.log(plugin);

  return {};
}
