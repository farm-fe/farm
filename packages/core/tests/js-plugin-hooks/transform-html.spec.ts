import { expect, test } from 'vitest';
import { getCompiler } from './common.js';

test('Js Plugin Execution - transformHtml', async () => {
  const hookName = 'transform-html';
  const calledHooks: string[] = [];
  const compiler = await getCompiler(
    '',
    [
      {
        name: 'test-transform-html-pre',
        transformHtml: {
          order: 0, // 0 means call this hook before parse and generate resources
          executor: async ({ htmlResource }) => {
            const code = Buffer.from(htmlResource.bytes as any).toString();
            calledHooks.push('transformHtmlPre');
            const replacedCode = code
              .replace('{head}', '<meta head />')
              .replace('{style}', '<meta style />')
              .replace('{ssr}', '<meta ssr />');
            return {
              ...htmlResource,
              bytes: [...Buffer.from(replacedCode)]
            };
          }
        }
      },
      {
        name: 'test-transform-html-post',
        transformHtml: {
          order: 2, // 2 means call this hook after parse and generate resources
          executor: async ({ htmlResource }) => {
            calledHooks.push('transformHtmlPost');
            const replacedCode = Buffer.from(htmlResource.bytes as any)
              .toString()
              .replace(/<meta (\w+?)>/g, '{$1}');
            htmlResource.bytes = [...Buffer.from(replacedCode)];
            return htmlResource;
          }
        }
      }
    ],
    hookName,
    {
      index: './index.html'
    },
    {
      entryFilename: 'index.html'
    }
  );

  await compiler.compile();

  expect(calledHooks).toEqual(['transformHtmlPre', 'transformHtmlPost']);

  const resourcesMap = compiler.resources();
  const html = resourcesMap['index.html'];
  expect(Buffer.from(html).toString()).matchSnapshot();
});
