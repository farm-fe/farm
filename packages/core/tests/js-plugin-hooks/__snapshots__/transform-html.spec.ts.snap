// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Js Plugin Execution - transformHtml 1`] = `
"<!doctype html><html lang=en><meta charset=UTF-8><meta name=viewport content="width=device-width,initial-scale=1.0">{head}{style}<title>Document</title><script data-farm-entry-script=true>
(globalThis || window || global)['f081367f80fe14896375a9f8b8918ca3'] = {};
(globalThis || window || global)['f081367f80fe14896375a9f8b8918ca3'] = {
  __FARM_TARGET_ENV__: 'browser',
};</script><script data-farm-entry-script=true>(function(r,e){var t={};function n(r){return Promise.resolve(o(r))}function o(e){if(t[e])return t[e].exports;var i={id:e,exports:{}};r[e](i,i.exports,o,n);t[e]=i;return i.exports}o(e)})({"433f8f8e":function  (e,t,n,u){"use strict";Object.defineProperty(t,"__esModule",{value:!0}),function(e,t){for(var n in t)Object.defineProperty(e,n,{enumerable:!0,get:t[n]});}(t,{ModuleSystem:function(){return f.ModuleSystem;},Plugin:function(){return l.FarmRuntimePlugin;}});let _=n("b15359af")._(n("52b7cefa")),f=n("ed7c2b04"),l=n("c30e104f");n("d9ce229e").__farm_global_this__.__farm_module_system__=(function(){let e=new f.ModuleSystem;return function(){return e;};})()(),(globalThis||window||global).f081367f80fe14896375a9f8b8918ca3.__farm_module_system__.setPlugins([_.default]);},"52b7cefa":function  (e,t,o,r){"use strict";Object.defineProperty(t,"__esModule",{value:!0}),Object.defineProperty(t,"default",{enumerable:!0,get:function(){return s;}});let n="undefined"!=typeof globalThis?globalThis:window,s={name:"farm-runtime-import-meta",_moduleSystem:{},bootstrap(e){this._moduleSystem=e;},moduleCreated(e){e.meta.env={mode:process.env.NODE_ENV,dev:"development"===process.env.NODE_ENV,prod:"production"===process.env.NODE_ENV};let t=this._moduleSystem.publicPaths?.[0]||"",{location:o}=n,r=o?\`\${o.protocol}//\${o.host}\${t.replace(/\\/$/,"")}/\${e.id}?t=\${Date.now()}\`:e.resource_pot;e.meta.url=r;}};},"966536f1":function  (e,t,r,n){"use strict";function o(e){if("function"!=typeof WeakMap)return null;var t=new WeakMap,r=new WeakMap;return(o=function(e){return e?r:t;})(e);}function u(e,t){if(!t&&e&&e.__esModule)return e;if(null===e||"object"!=typeof e&&"function"!=typeof e)return{default:e};var r=o(t);if(r&&r.has(e))return r.get(e);var n={__proto__:null},u=Object.defineProperty&&Object.getOwnPropertyDescriptor;for(var f in e)if("default"!==f&&Object.prototype.hasOwnProperty.call(e,f)){var i=u?Object.getOwnPropertyDescriptor(e,f):null;i&&(i.get||i.set)?Object.defineProperty(n,f,i):n[f]=e[f];}return n.default=e,r&&r.set(e,n),n;}Object.defineProperty(t,"__esModule",{value:!0}),function(e,t){for(var r in t)Object.defineProperty(e,r,{enumerable:!0,get:t[r]});}(t,{_:function(){return u;},_interop_require_wildcard:function(){return u;}});},"b15359af":function  (e,n,t,r){"use strict";function u(e){return e&&e.__esModule?e:{default:e};}Object.defineProperty(n,"__esModule",{value:!0}),function(e,n){for(var t in n)Object.defineProperty(e,t,{enumerable:!0,get:n[t]});}(n,{_:function(){return u;},_interop_require_default:function(){return u;}});},"c30e104f":function  (e,t,i,l){"use strict";Object.defineProperty(t,"__esModule",{value:!0}),Object.defineProperty(t,"FarmRuntimePluginContainer",{enumerable:!0,get:function(){return n;}});class n{plugins=[];constructor(e){this.plugins=e;}hookSerial(e,...t){for(let i of this.plugins){let l=i[e];l&&l.apply(i,t);}}hookBail(e,...t){for(let i of this.plugins){let l=i[e];if(l){let e=l.apply(i,t);if(e)return e;}}}}},"d2805128":function  (e,t,r,i){"use strict";Object.defineProperty(t,"__esModule",{value:!0}),Object.defineProperty(t,"Module",{enumerable:!0,get:function(){return s;}});class s{id;exports;resource_pot;meta;require;constructor(e,t){this.id=e,this.exports={},this.meta={env:{}},this.require=t;}}},"d9ce229e":function  (module,exports,farmRequire,farmDynamicRequire){"use strict";function _export(e,t){for(var r in t)Object.defineProperty(e,r,{enumerable:!0,get:t[r]});}Object.defineProperty(exports,"__esModule",{value:!0}),_export(exports,{ResourceLoader:function(){return ResourceLoader;},__farm_global_this__:function(){return __farm_global_this__;},isBrowser:function(){return isBrowser;},targetEnv:function(){return targetEnv;}});let __farm_global_this__=eval("(globalThis || window || global)['f081367f80fe14896375a9f8b8918ca3']"),targetEnv=__farm_global_this__.__FARM_TARGET_ENV__||"node",isBrowser="browser"===targetEnv&&(globalThis||window).document;class ResourceLoader{moduleSystem;_loadedResources;_loadingResources;publicPaths;constructor(e,t){this.moduleSystem=e,this._loadedResources={},this._loadingResources={},this.publicPaths=t;}load(e,t=0){if(!isBrowser){let t=this.moduleSystem.pluginContainer.hookBail("loadResource",e);if(t)return t.then(t=>{if(!t.success&&t.retryWithDefaultResourceLoader){if("script"===e.type)return this._loadScript(\`./\${e.path}\`);if("link"===e.type)return this._loadLink(\`./\${e.path}\`);}else if(!t.success)throw Error(\`[Farm] Failed to load resource: "\${e.path}, type: \${e.type}". Original Error: \${t.err}\`);});if("script"===e.type)return this._loadScript(\`./\${e.path}\`);if("link"===e.type)return this._loadLink(\`./\${e.path}\`);}let r=this.publicPaths[t],o=\`\${r.endsWith("/")?r.slice(0,-1):r}/\${e.path}\`;if(this._loadedResources[e.path])return;if(this._loadingResources[e.path])return this._loadingResources[e.path];let s=this.moduleSystem.pluginContainer.hookBail("loadResource",e);return s?s.then(r=>{if(r.success)this.setLoadedResource(e.path);else if(r.retryWithDefaultResourceLoader)return this._load(o,e,t);else throw Error(\`[Farm] Failed to load resource: "\${e.path}, type: \${e.type}". Original Error: \${r.err}\`);}):this._load(o,e,t);}setLoadedResource(e){this._loadedResources[e]=!0;}isResourceLoaded(e){return this._loadedResources[e];}_load(e,t,r){let o=Promise.resolve();return"script"===t.type?o=this._loadScript(e):"link"===t.type&&(o=this._loadLink(e)),this._loadingResources[t.path]=o,o.then(()=>{this._loadedResources[t.path]=!0;}).catch(o=>{if(console.warn(\`[Farm] Failed to load resource "\${e}" using publicPath: \${this.publicPaths[r]}\`),++r<this.publicPaths.length)return this._load(e,t,r);throw Error(\`[Farm] Failed to load resource: "\${t.path}, type: \${t.type}". \${o}\`);}),o;}_loadScript(e){return isBrowser?new Promise((t,r)=>{let o=document.createElement("script");o.src=e,document.body.appendChild(o),o.onload=()=>{t();},o.onerror=e=>{r(e);};}):import(e);}_loadLink(e){return isBrowser?new Promise((t,r)=>{let o=document.createElement("link");o.rel="stylesheet",o.href=e,document.head.appendChild(o),o.onload=()=>{t();},o.onerror=e=>{r(e);};}):Promise.resolve();}}},"ed3f3d30":function  (e,t,n,r){"use strict";function o(e,t){return Object.keys(e).forEach(function(n){"default"===n||Object.prototype.hasOwnProperty.call(t,n)||Object.defineProperty(t,n,{enumerable:!0,get:function(){return e[n];}});}),e;}Object.defineProperty(t,"__esModule",{value:!0}),function(e,t){for(var n in t)Object.defineProperty(e,n,{enumerable:!0,get:t[n]});}(t,{_:function(){return o;},_export_star:function(){return o;}});},"ed7c2b04":function  (e,t,r,i){"use strict";Object.defineProperty(t,"__esModule",{value:!0}),Object.defineProperty(t,"ModuleSystem",{enumerable:!0,get:function(){return p;}});let s=r("d2805128"),o=r("c30e104f"),l=r("d9ce229e"),n=r("b15359af"),u=r("966536f1"),h=r("ed3f3d30"),d=globalThis||window,c={"@swc/helpers/_/_interop_require_default":{default:n._interop_require_default,_:n._interop_require_default},"@swc/helpers/_/_interop_require_wildcard":{default:u._interop_require_wildcard,_:u._interop_require_wildcard},"@swc/helpers/_/_export_star":{default:h._export_star,_:h._export_star}};class p{modules;cache;externalModules;publicPaths;dynamicModuleResourcesMap;resourceLoader;pluginContainer;targetEnv;constructor(){this.modules={},this.cache={},this.publicPaths=[],this.dynamicModuleResourcesMap={},this.resourceLoader=new l.ResourceLoader(this,this.publicPaths),this.pluginContainer=new o.FarmRuntimePluginContainer([]),this.targetEnv=l.targetEnv,this.externalModules={};}require(e,t=!1){if(c[e])return c[e];if(this.cache[e]&&!this.pluginContainer.hookBail("readModuleCache",this.cache[e]))return this.cache[e].exports;let r=this.modules[e];if(!r){if(this.externalModules[e]){let r=this.externalModules[e];return t&&r.default||r;}return("node"===this.targetEnv||!l.isBrowser)&&nodeRequire?nodeRequire(e):(this.pluginContainer.hookSerial("moduleNotFound",e),console.debug(\`[Farm] Module "\${e}" is not registered\`),{});}let i=new s.Module(e,this.require.bind(this));i.resource_pot=r.__farm_resource_pot__,this.pluginContainer.hookSerial("moduleCreated",i),this.cache[e]=i,(globalThis||global||window||{}).require||((globalThis||global||window||{require:undefined}).require=this.require.bind(this));let o=r(i,i.exports,this.require.bind(this),this.farmDynamicRequire.bind(this));return o&&o.then?o.then(()=>(this.pluginContainer.hookSerial("moduleInitialized",i),i.exports)):(this.pluginContainer.hookSerial("moduleInitialized",i),i.exports);}farmDynamicRequire(e){if(this.modules[e]){let t=this.require(e);return t.__farm_async?t.default:Promise.resolve(t);}let t=this.dynamicModuleResourcesMap[e];if(!t||0===t.length)throw Error(\`Dynamic imported module "\${e}" does not belong to any resource\`);return Promise.all(t.map(e=>this.resourceLoader.load(e))).then(()=>{if(!this.modules[e])throw Error(\`Dynamic imported module "\${e}" is not registered.\`);let t=this.require(e);return t.__farm_async?t.default:t;}).catch(t=>{throw console.error(\`[Farm] Error loading dynamic module "\${e}"\`,t),t;});}register(e,t){if(this.modules[e]){console.warn(\`Module "\${e}" has registered! It should not be registered twice\`);return;}this.modules[e]=t;}update(e,t){this.modules[e]=t,this.clearCache(e);}delete(e){return!!this.modules[e]&&(this.clearCache(e),delete this.modules[e],!0);}getModuleUrl(e){let t=this.publicPaths[0]??"";return d.location?\`\${d.location.protocol}//\${d.location.host}\${t.endsWith("/")?t.slice(0,-1):t}/\${this.modules[e].__farm_resource_pot__}\`:this.modules[e].__farm_resource_pot__;}getCache(e){return this.cache[e];}clearCache(e){return!!this.cache[e]&&(delete this.cache[e],!0);}setInitialLoadedResources(e){for(let t of e)this.resourceLoader.setLoadedResource(t);}setDynamicModuleResourcesMap(e){this.dynamicModuleResourcesMap=e;}setPublicPaths(e){this.publicPaths=e,this.resourceLoader.publicPaths=this.publicPaths;}setPlugins(e){this.pluginContainer.plugins=e;}addPlugin(e){this.pluginContainer.plugins.every(t=>t.name!==e.name)&&this.pluginContainer.plugins.push(e);}removePlugin(e){this.pluginContainer.plugins=this.pluginContainer.plugins.filter(t=>t.name!==e);}setExternalModules(e){Object.assign(this.externalModules,e||{});}bootstrap(){this.pluginContainer.hookSerial("bootstrap",this);}}},},"433f8f8e");</script><div id=app-container>{ssr}</div><script data-farm-entry-script=true>(globalThis || window || global)['f081367f80fe14896375a9f8b8918ca3'].__farm_module_system__.setInitialLoadedResources([]);</script><script data-farm-entry-script=true>(globalThis || window || global)['f081367f80fe14896375a9f8b8918ca3'].__farm_module_system__.setDynamicModuleResourcesMap({  });</script><script data-farm-entry-script=true>(globalThis || window || global)['f081367f80fe14896375a9f8b8918ca3'].__farm_module_system__.setPublicPaths(['/']);</script><script data-farm-entry-script=true>(globalThis || window || global)['f081367f80fe14896375a9f8b8918ca3'].__farm_module_system__.bootstrap();</script>"
`;
